### PR TITLE
RFC: Sphinx setup & documentation

### PR DESCRIPTION
**Pull request by agger-magenta**
The framework is there - is this more or less the level of documentation we're aiming for?

Next step is to put it on Read The Docs, but that's for when it's on the development branch.

### CODE REVIEW

File: backend/docs/development_docs/api_requests_per_screen.rst [hunk 1]
```diff
@@ -70,54 +76,109 @@ SOLR will return data in JSON format and according to the following template: ::
...
+        "_version_":1617934905120915456}    
+    ]
      }}
 
+In this output, the fields ``id`` and ``_version_`` are not used. The
+actual contents of the found object are in the ``document`` field, which
+is a JSON encoded string. 
+
+If the ``document`` field of any of these results is parsed, we may, e.g., get an organisation
+unit that looks like this: ::
```
Comment (danchr) [inline backend/docs/development_docs/api_requests_per_screen.rst]: `s/: ::/::/g`
```suggestion
unit that looks like this::
```

Also, iff the source is valid JSON, you should mark it as such.

File: backend/docs/development_docs/api_requests_per_screen.rst [hunk 2]
```diff
@@ -70,54 +76,109 @@ SOLR will return data in JSON format and according to the following template: ::
...
-          "Problemknuser",
-          "Specialkonsulent"],
-        "managers":["Direktør",
-          "Esther Hildeborg Nielsen",
-          "b078107a-4299-4251-b41c-23e0b8a9575a"],
-        "id":"38bbb6ad-d31f-4381-aa88-61be37f70d8c",
-        "_version_":1616659562091249664}]
+        {
+           "document":"<very long string omitted>",
+        "id":"8284b4b5-118a-4cbf-aa7c-f08ae2ca391c",
```
Comment (danchr) [inline backend/docs/development_docs/api_requests_per_screen.rst]: Indentation is weird. Also, should probably be styled as JSON, assuming you edit it to be valid.

File: backend/docs/development_docs/api_requests_per_screen.rst [hunk 3]
```diff
@@ -70,54 +76,109 @@ SOLR will return data in JSON format and according to the following template: ::
...
-dislayed is simply "Hjørring".  The members are, apart from the obvious
-ones, to be interpreted like this:
+dislayed is "Budget og Planlægning".  The members are, apart from the
+obvious ones, to be interpreted like this:
 
 * ``locations`` are addresses and can be of type ``DAR``, ``EMAIL``
-  and ``PHONE``. The list contains a list of addresses that are to be
-  interpreted as couples, *(type, value)*.
+  and ``PHONE``. The list contains a list of addresses; each address is
+  itself a couple,  *[type, value]*.
```
Comment (danchr) [inline backend/docs/development_docs/api_requests_per_screen.rst]: I'm fairly certain you mean a [pair](https://www.merriam-webster.com/dictionary/pair), not a [couple](https://www.merriam-webster.com/dictionary/couple). (Unless the type really is married to the value? 😁) Strictly speaking, the addresses aren't two-tuples or pairs — given that neither exists in the JSON realm — but rather two-element lists or arrays.

Furthermore, you should probably expand on what _`type`_ and _`value`_ are.

File: backend/docs/development_docs/api_requests_per_screen.rst [hunk 4]
```diff
@@ -128,25 +189,25 @@ Employees
 .........
 
 The result headers are as in the example above - only the contents of
-the "docs" section are different.. A typical "doc" for a person could
+the "document" field are different.. A typical "document" for a person could
```
Comment (danchr) [inline backend/docs/development_docs/api_requests_per_screen.rst]:
```suggestion
the ``document`` field are different.. A typical ``document`` for a person could
```

File: backend/docs/development_docs/api_requests_per_screen.rst [hunk 5]
```diff
@@ -128,25 +189,25 @@ Employees
...
         "uuid":["cee8800a-983d-41fa-998c-b4557d68ec35"],
         "name":["Anna Bjerre Reidl"],
-        "locations":["PHONE",
-          "21557342",
-          "DAR",
-          "Strandgårdsvej 16, Skallerup Klit, 9800 Hjørring",
-          "EMAIL",
-          "annar@hjorring.dk"],
-        "departments":["IT-Support",
+        "locations":[["PHONE",
```
Comment (danchr) [inline backend/docs/development_docs/api_requests_per_screen.rst]: This chunk should be highlighted as JSON.

File: backend/import/do_import.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+"""Import org units and employees from MO."""
```
Comment (danchr) [inline backend/import/do_import.py]: The doc string normally goes _below_ any leading comment and imports IMO.

File: backend/import/do_import.py [hunk 7]
```diff
@@ -63,8 +64,7 @@ def get_orgunit_data(ou):
 
 
 def get_employee_data(employee):
-    '''Get the data we need to display this employee'''
-
+    """Get the data we need to display this employee."""
```
Comment (danchr) [inline backend/import/do_import.py]: That's mostly in the function name, isn't it?

File: backend/import/do_import.py [hunk 8]
```diff
@@ -138,7 +138,7 @@ def handle_this(uuid):
 
 
 def file_writer(directory, field_name='uuid'):
-
+    """Return a function that writes data to a given directory."""
```
Comment (danchr) [inline backend/import/do_import.py]: How?

File: backend/import/mo_api.py [hunk 9]
```diff
@@ -1,3 +1,4 @@
+"""A simple API for requesting data from MO."""
```
Comment (danchr) [inline backend/import/mo_api.py]: See earlier comment.

File: backend/import/mo_api.py [hunk 10]
```diff
@@ -40,17 +42,23 @@ class MOData:
     """Abstract base class to interface with MO objects."""
 
     def __init__(self, uuid):
+        """Initialize object - speccify session and getter method."""
```
Comment (danchr) [inline backend/import/mo_api.py]:
```suggestion
        """Initialize object - specify session and getter method."""
```

Neither `session` nor `getter` appear as arguments to this function, which is what the docstring should describe. Perhaps a comment instead? What's `uuid`, though?

File: docs/source/backend/architecture.rst [hunk 11]
```diff
@@ -0,0 +1,18 @@
+Architecture
+============
+
+The os2phonebook backend consists of two components:
+
+* A Python program (located in the ``import`` directory) that imports
+  all the necessary data from MO and stores it in the file system, and
+* An SOLR server in which the data are indexed, and from which the
```
Comment (danchr) [inline docs/source/backend/architecture.rst]: I believe “SOLR” is pronounced “solar” 🙂
```suggestion
* A SOLR server in which the data are indexed, and from which the
```

File: docs/source/backend/architecture.rst [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+  frontend retrieves them.
+
+This corresponds to the two steps in the ``import_data.sh`` script.
+
+The phonebook is supposed to be kept up to date by importing and
+reindexing regularly, normally once a day. When reindexing, the SOLR
+data are completely removed and reimported, which means that data may be
```
Comment (danchr) [inline docs/source/backend/architecture.rst]: Data is singular — perhaps “the data in SOLR is”?

File: docs/source/backend/installation.rst [hunk 13]
```diff
@@ -0,0 +1,50 @@
+
+Installation
+============
+
+To check out the source code and quickly install the backend, execute
+the following on a computer with a modern version of GNU/Linux::
```
Comment (danchr) [inline docs/source/backend/installation.rst]: Heh 🙂

File: docs/source/backend/do_import.rst [hunk 14]
```diff
@@ -0,0 +1,7 @@
+do\_import module
+=================
+
+.. automodule:: do_import
```
Comment (danchr) [inline docs/source/backend/do_import.rst]: This page is [empty](https://os2mo.readthedocs.io/projects/phonebook/en/feature-25304_documentation_and_sphinx/backend/do_import.html)?

File: backend/docs/development_docs/api_requests_per_screen.rst (hunk 1 above)
Comment (agger-magenta) [reply backend/docs/development_docs/api_requests_per_screen.rst]: The `.../g` indicates that all occurences on each line should be replaced, not just the first, so hopefully it won't be needed. Apart from that, I follow your suggestion!

File: backend/docs/development_docs/api_requests_per_screen.rst (hunk 2 above)
Comment (agger-magenta) [reply backend/docs/development_docs/api_requests_per_screen.rst]: Will indent & mark as JSON.

File: backend/import/do_import.py (hunk 6 above)
Comment (agger-magenta) [reply backend/import/do_import.py]: PEP 257 says that a docstring "is a string literal that occurs as the first statement in a module, function, class, or method definition".

As imports are statements, this means it should go _before_ those. However, it's not a problem to put it after the license boilerplate.

File: backend/import/do_import.py (hunk 7 above)
Comment (agger-magenta) [reply backend/import/do_import.py]: You propose omitting this docstring, then?

Comment (danchr) [reply backend/import/do_import.py]: TBH I'm just complaining — I suppose something is better than nothing, but what's there doesn't really add much ;)

File: backend/docs/development_docs/api_requests_per_screen.rst (hunk 3 above)
Comment (agger-magenta) [reply backend/docs/development_docs/api_requests_per_screen.rst]: Yes, done!

File: backend/docs/development_docs/api_requests_per_screen.rst (hunk 4 above)
Comment (agger-magenta) [reply backend/docs/development_docs/api_requests_per_screen.rst]: :+1:

File: backend/docs/development_docs/api_requests_per_screen.rst (hunk 5 above)
Comment (agger-magenta) [reply backend/docs/development_docs/api_requests_per_screen.rst]: :+1:

File: backend/import/do_import.py (hunk 8 above)
Comment (agger-magenta) [reply backend/import/do_import.py]: Added explanation.

File: backend/import/mo_api.py (hunk 9 above)
Comment (agger-magenta) [reply backend/import/mo_api.py]: :+1:

File: backend/import/mo_api.py (hunk 10 above)
Comment (agger-magenta) [reply backend/import/mo_api.py]: Fixed.

File: docs/source/backend/architecture.rst (hunk 11 above)
Comment (agger-magenta) [reply docs/source/backend/architecture.rst]: :+1:

File: docs/source/backend/architecture.rst (hunk 12 above)
Comment (agger-magenta) [reply docs/source/backend/architecture.rst]: That is [up for discussion](https://www.theguardian.com/news/datablog/2010/jul/16/data-plural-singular) among Anglophones. I'll leave it as is until they agree. :)

File: docs/source/backend/do_import.rst (hunk 14 above)
Comment (agger-magenta) [reply docs/source/backend/do_import.rst]: No longer!

**Pull request merged**